lib/block_view_lockups.go

```diff
@@ -1695,6 +2506,53 @@ func (bav *UtxoView) _connectCoinUnlock(
 		bav._deleteLockedBalanceEntry(unlockableLockedBalanceEntry)
 	}
 
+	// Unlock all vested locked balance entries.
+	// NOTE: See the comment on LockedBalanceEntryKey for how we deal with modified vested locked balance entries.
+	var prevVestedLockedBalanceEntries []*LockedBalanceEntry
+	for _, unlockableLockedBalanceEntry := range vestedUnlockableLockedBalanceEntries {
+		// Depending on the time of unlock, compute how much from the balance can be unlocked.
+		amountToUnlock, err := CalculateVestedEarnings(unlockableLockedBalanceEntry, blockTimestampNanoSecs)
+		if err != nil {
+			return 0, 0, nil,
+				errors.Wrap(err, "_connectCoinUnlock: error computing vested earnings")
+		}
+
+		// Add the unlocked amount and check for overflow.
+		unlockedBalance, err =
+			SafeUint256().Add(unlockedBalance, amountToUnlock)
+		if err != nil {
+			return 0, 0, nil,
+				errors.Wrap(RuleErrorCoinUnlockUnlockableCoinsOverflow, "_connectCoinUnlock")
+		}
+
+		// Append the original LockedBalanceEntry in the event we rollback the transaction.
+		prevVestedLockedBalanceEntries = append(prevVestedLockedBalanceEntries, unlockableLockedBalanceEntry.Copy())
+
+		// Depending on when the unlock occurs, we either DELETE or MODIFY the locked balance entry.
+		if blockTimestampNanoSecs >= unlockableLockedBalanceEntry.VestingEndTimestampNanoSecs {
+			bav._deleteLockedBalanceEntry(unlockableLockedBalanceEntry)
+		} else {
+			// DELETE the previous key.
+			bav._deleteLockedBalanceEntry(unlockableLockedBalanceEntry)
+
+			// Create and modify a copy to prevent pointer reuse.
+			modifiedLockedBalanceEntry := unlockableLockedBalanceEntry.Copy()
+			modifiedLockedBalanceEntry.UnlockTimestampNanoSecs = blockTimestampNanoSecs
```

Comment:
I was noodling trying to figure out if this should be `blockTimestampNanoSecs+1` instead. I think it's right as-is but noodle on it.